### PR TITLE
Update documentation about enabling sanitizers

### PR DESCRIPTION
**Pull request by JanuszL**
- add an ability to enable ubsan, and lsan during DALI build
- updates the guide on how to run the sanitizers and workaround
  the issue described in https://github.com/google/sanitizers/issues/934
- fixes memory leak in ArgumentInst

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Description
- [x] **Bug fix** (*non-breaking change which fixes an issue*)
- [x] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
- add an ability to enable ubsan, and lsan during DALI build
- updates the guide on how to run the sanitizers and workaround
  the issue described in https://github.com/google/sanitizers/issues/934
- fixes memory leak in ArgumentInst

#### Additional information
- Affected modules and functionalities:
  - documentation
  - build system
  - argument.cc and argument.h

- Key points relevant for the review:
  - should it be split into two PRs?

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [x] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2317

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

File: dali/pipeline/data/tensor_test.cc [hunk 1]
```diff
@@ -436,7 +436,7 @@ TYPED_TEST(TensorTest, TestCustomAlloc) {
   std::function<void(void *)> deleter;
   if (std::is_same<TypeParam, CPUBackend>::value) {
     alloc_func = [&](size_t bytes) {
-      return new uint8_t[bytes];
+      return malloc(sizeof(uint8_t) * bytes);
```
Comment (banasraf) [inline dali/pipeline/data/tensor_test.cc:439]: Why malloc instead of new?

Comment (JanuszL) [reply dali/pipeline/data/tensor_test.cc:439]: Because we use plain `free` in L442. I can use `delete []` there instead.

**Review by awolant — APPROVED** [review]

**Comment by JanuszL** [comment]
!build

**Review by banasraf — APPROVED** [review]

**Pull request merged**